coders/qoi.c

```diff
@@ -0,0 +1,450 @@
+  if (status == MagickFalse)
+    return(DestroyImageList(image));
+  quantum_info=AcquireQuantumInfo(image_info,image);
+  if (quantum_info == (QuantumInfo *) NULL)
+    ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
+  status=SetQuantumFormat(image,quantum_info,UnsignedQuantumFormat);
+  if (status == MagickFalse)
+    return(DestroyImageList(image));
+
+  /*
+    Get a write pointer for the whole image.
+  */
+  q=QueueAuthenticPixels(image,0,0,image->columns,image->rows,exception);
+  if (q == (Quantum *) NULL)
+    ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
+  /*
+    Initialize decoding state.
+  */
+  px.rgba.r=0;
+  px.rgba.g=0;
+  px.rgba.b=0;
+  px.rgba.a=255;
+  memset(lut,0,sizeof(lut));
+  /*
+    Actual decoding.
+  */
+  for (p=0; p < (image->rows * image->columns);)
+  {
+    run = 0;
+
+    if ((b=ReadBlobByte(image)) == EOF)
+      break;
+    if (b == QOI_OP_RGB) {
+      if ((b=ReadBlobByte(image)) == EOF)
+        break;
+      px.rgba.r=b;
+      if ((b=ReadBlobByte(image)) == EOF)
+        break;
+      px.rgba.g=b;
+      if ((b=ReadBlobByte(image)) == EOF)
+        break;
+      px.rgba.b=b;
+    }
+    else if (b == QOI_OP_RGBA) {
+      if ((b=ReadBlobByte(image)) == EOF)
+        break;
+      px.rgba.r=b;
+      if ((b=ReadBlobByte(image)) == EOF)
+        break;
+      px.rgba.g=b;
+      if ((b=ReadBlobByte(image)) == EOF)
+        break;
+      px.rgba.b=b;
+      if ((b=ReadBlobByte(image)) == EOF)
+        break;
+      px.rgba.a=b;
+    }
+    else if ((b & QOI_MASK_2) == QOI_OP_INDEX) {
+      px=lut[b & ~QOI_MASK_2];
+    }
+    else if ((b & QOI_MASK_2) == QOI_OP_DIFF) {
+      px.rgba.r+=((b >> 4) & 0x03) - 2;
+      px.rgba.g+=((b >> 2) & 0x03) - 2;
+      px.rgba.b+=( b       & 0x03) - 2;
+    }
+    else if ((b & QOI_MASK_2) == QOI_OP_LUMA) {
+      vg=(b & ~QOI_MASK_2) - 32;
+      if ((b=ReadBlobByte(image)) == EOF)
+        break;
+      px.rgba.r+=vg - 8 + ((b >> 4) & 0x0f);
+      px.rgba.g+=vg;
+      px.rgba.b+=vg - 8 +  (b       & 0x0f);
+    }
+    else if ((b & QOI_MASK_2) == QOI_OP_RUN) {
+      run=b & ~QOI_MASK_2;
+    }
+    lut[QOI_COLOR_HASH(px) % 64]=px;
+    do {
+      SetPixelRed(image,ScaleCharToQuantum((unsigned char)px.rgba.r),q);
+      SetPixelGreen(image,ScaleCharToQuantum((unsigned char)px.rgba.g),q);
+      SetPixelBlue(image,ScaleCharToQuantum((unsigned char)px.rgba.b),q);
+      if (channels == 4)
+        SetPixelAlpha(image,ScaleCharToQuantum((unsigned char)px.rgba.a),q);
+      q+=GetPixelChannels(image);
+      p++;
+    } while (run-- > 0);
+    status=SetImageProgress(image,LoadImageTag,p,image->rows * image->columns);
+    if (status == MagickFalse)
+      break;
+  }
+  if (EOFBlob(image) != MagickFalse)
+    ThrowFileException(exception,CorruptImageError,
+      "UnexpectedEndOfFile",image->filename);
+  if (status != MagickFalse) {
+    status=SyncAuthenticPixels(image,exception);
+    if (image->ping == MagickFalse)
```

Comment:
We can do an ealier exit when we only are pinging the image?

Reply:
I didn't understood what ping mean, I've copied from other coders... I don't understand what you are suggesting sorry.

Reply:
`ping` means that we only want the metadata of the image. So you can do something like this before you read the pixels:

```
if (image->ping != MagickFalse)
  return(GetFirstImageInList(image))
```